feature: add rule to refactor exception methods

alfredbez opened the pull request:
see: https://github.com/sebastianbergmann/phpunit/issues/3775

alfredbez commented:
Is it possible to ensure that we're not breaking expected behavior or should I just remove the fixture?

TomasVotruba commented:
CI needs fixing

alfredbez commented:
I know, my question was just about the failing step

alfredbez commented:
Maybe it makes also sense to include the CI checks to the `complete-check` script in composer, e.g.:
```diff
git --no-pager diff
diff --git a/composer.json b/composer.json
index 826e67b1c..d8494b790 100644
--- a/composer.json
+++ b/composer.json
@@ -213,11 +213,13 @@
     "scripts": {
         "complete-check": [
             "@check-cs",
+            "@check-fixtures",
             "phpunit",
             "@phpstan",
             "@docs"
         ],
         "check-cs": "ecs check bin packages src tests utils compiler --ansi",
+        "check-fixtures": "php ci/check_keep_fixtures.php",
         "fix-cs": [
             "ecs check bin packages src tests utils compiler --fix --ansi",
             "ci/clean_trailing_spaces.sh"
```
Fast feedback is really helpful

TomasVotruba commented:
> Maybe it makes also sense to include the CI checks to the complete-check script in composer, e.g.:

Yes, why not

TomasVotruba commented on packages/PHPUnit/tests/Rector/MethodCall/ExplicitPhpErrorApiRector/Fixture/php_combination.php.inc:
```diff
@@ -0,0 +1,37 @@
+        $this->expectException(\PHPUnit\Framework\TestCase\Warning::class);
+    }
+}
+
```
This class duplicates the other 4 cases right?
What's the added value?

alfredbez replied:
fixed that

TomasVotruba commented:
> Is it possible to ensure that we're not breaking expected behavior or should I just remove the fixture?

I don't know what you mean.

alfredbez commented:
> > Maybe it makes also sense to include the CI checks to the complete-check script in composer, e.g.:
> 
> Yes, why not

I'll create another PR for that. I rebased this PR again and fixed the fixture, ready for review.

alfredbez commented on packages/PHPUnit/src/Rector/MethodCall/ExplicitPhpErrorApiRector.php:
```diff
@@ -0,0 +1,102 @@
+        if (isset($node->args[0]) && property_exists(
+            $node->args[0]->value,
+            'class'
+        ) && (string) $node->args[0]->value->class === $exceptionClass) {
```
I don't like this part, especially the `property_exists` check, but I'm not sure if there is a better way, to check if `$this->expectException()` has an argument that matches e.g. `PHPUnit\Framework\TestCase\Warning`

TomasVotruba replied:
Let's merge if for now. 
I'll look on this as part of common rule cleanup.

TomasVotruba commented:
Thank you :+1:
